service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,318 @@
+	return Arguments{
+		ID:               agentseed.Get().UID,
+		Metadata:         make(map[string]string),
+		PollFrequency:    1 * time.Minute,
+		HTTPClientConfig: config.CloneDefaultHTTPClientConfig(),
+	}
+}
+
+// SetToDefault implements river.Defaulter.
+func (a *Arguments) SetToDefault() {
+	*a = GetDefaultArguments()
+}
+
+// Validate implements river.Validator.
+func (a *Arguments) Validate() error {
+	// We must explicitly Validate because HTTPClientConfig is squashed and it
+	// won't run otherwise
+	if a.HTTPClientConfig != nil {
+		return a.HTTPClientConfig.Validate()
+	}
+
+	return nil
+}
+
+// Hash marshals the Arguments and returns a hash representation.
+func (a *Arguments) Hash() (string, error) {
+	b, err := river.Marshal(a)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal arguments: %w", err)
+	}
+	return getHash(b), nil
+}
+
+// New returns a new instance of the remotecfg service.
+func New(opts Options) (*Service, error) {
+	basePath := filepath.Join(opts.StoragePath, ServiceName)
+	err := os.MkdirAll(basePath, 0750)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		opts:   opts,
+		ticker: time.NewTicker(math.MaxInt64),
+	}, nil
+}
+
+// Data is a no-op for the remotecfg service.
+func (s *Service) Data() any {
+	return nil
+}
+
+// Definition returns the definition of the remotecfg service.
+func (s *Service) Definition() service.Definition {
+	return service.Definition{
+		Name:       ServiceName,
+		ConfigType: Arguments{},
+		DependsOn:  nil, // remotecfg has no dependencies.
+	}
+}
+
+var _ service.Service = (*Service)(nil)
+
+// Run implements [service.Service] and starts the remotecfg service. It will
+// run until the provided context is canceled or there is a fatal error.
+func (s *Service) Run(ctx context.Context, host service.Host) error {
+	s.ctrl = host.NewController(ServiceName)
+
+	s.fetch()
+
+	// Run the service's own controller.
+	go func() {
+		s.ctrl.Run(ctx)
+	}()
+
+	for {
+		select {
+		case <-s.ch:
+			b, err := s.getAPIConfig()
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to fetch configuration from the API", "err", err)
+				continue
+			}
+
+			// The polling loop got the same configuration, no need to reload.
+			newConfigHash := getHash(b)
+			if s.getCfgHash() == newConfigHash {
+				level.Debug(s.opts.Logger).Log("msg", "skipping to the next polling loop")
+				continue
+			}
+
+			// We have a new configuration let's try to load it
+			err = s.parseAndLoad(b)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to load configuration from the API", "err", err)
+				continue
+			}
+
+			level.Info(s.opts.Logger).Log("msg", "new remote configuration loaded successfully")
+
+			// If successful, flush to disk and keep a copy.
+			err = os.WriteFile(s.dataPath, b, 0750)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to flush remote configuration contents the on-disk cache", "err", err)
```

Comment:
To be clear we are fine with the configuration it is running being different from the value on the filesystem?

Reply:
Sorry, I didn't get what you mean here.

Reply:
If it fails writing the new config it will still load the remote config so now the cache and running configs are different. The only big use case I could see is if we needed them to investigate the cache, but I think that is pretty far fetched.

Reply:
Yeah, I don't see this as a major issue TBH. The cache is a best-effort thing only, the service should run fine even if storage-path is not writeable.